rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+        &mut out_size as *mut _,
+        &mut out_array as *mut _
+    ));
+    let list = unsafe { slice::from_raw_parts(out_array, out_size as usize) };
+    let list = list
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_bytes())
+        .map(|bs| str::from_utf8(bs).unwrap())
+        .collect();
+    Mutex::new(list)
+}
+
+/// Returns a registered TVM function by name.
+pub fn get_global_func(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+    let name = CString::new(name).unwrap();
+    let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+    check_call!(ts::TVMFuncGetGlobal(
+        name.as_ptr() as *const c_char,
+        &mut handle as *mut _
+    ));
+    if !(handle.is_null()) {
+        mem::forget(name);
+        return Some(Function::new(handle, is_global, false));
+    } else {
+        if allow_missing {
+            return None;
+        } else {
+            panic!("Cannot find global function {:?}", name);
+        }
+    }
+}
+
+/// Wrapper around TVM function handle.
+#[derive(Debug, Hash)]
+pub struct Function {
+    handle: ts::TVMFunctionHandle,
+    is_global: bool,
+    is_released: bool,
+}
+
+impl Function {
+    pub(crate) fn new(handle: ts::TVMFunctionHandle, is_global: bool, is_released: bool) -> Self {
+        Function {
+            handle: handle,
+            is_global: is_global,
+            is_released: is_released,
+        }
+    }
+
+    /// For a given function, it returns a function by name.
+    pub fn get_function(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+        let name = CString::new(name).unwrap();
+        GLOBAL_FUNCTION_NAMES
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|&&s| s == name.to_str().expect("Invalid UTF-8 name"))
+            .map(|nm| get_global_func(nm, is_global, allow_missing).unwrap())
+    }
+
+    /// Returns the underlying TVM function handle.
+    pub fn as_handle(&self) -> ts::TVMFunctionHandle {
+        self.handle
+    }
+
+    pub fn is_global(&self) -> bool {
+        self.is_global
+    }
+
+    pub fn is_released(&self) -> bool {
+        self.is_released
+    }
+}
+
+impl Clone for Function {
+    fn clone(&self) -> Function {
+        if !self.is_released {
+            Self {
+                handle: self.handle,
+                is_global: true,
+                is_released: false,
+            }
+        } else {
+            panic!("Released function cannot be cloned!");
+        }
+    }
+}
+
+impl Drop for Function {
+    fn drop(&mut self) {
+        if !self.is_released & !self.is_global {
+            check_call!(ts::TVMFuncFree(self.handle));
+            self.is_released = true;
+        }
+    }
+}
+
+/// Function builder in order to create and call functions.
+/// *Note:* TVM only accepts at most one return value.
+#[derive(Debug, Clone, Default)]
+pub struct Builder<'a> {
```

Comment:
suggestion: replace this with a simple `#[derive(Builder)]` atop `struct Function` using the [derive-builder](https://docs.rs/derive_builder/0.7.0/derive_builder/) crate.

Reply:
Thanks! I'll look into it after this PR. There're some restrictions like there can be at most one return val for packed functions that I need to figure out how to configure the derive.